### ISSUE TITLE
Update companion container that updates Mathesar container on the fly

### ISSUE THREAD

**Issue by silentninja** [centerofci/mathesar#2072]
Implement the equivalent of the [Wiki.js Update Companion](https://github.com/Requarks/wiki-update-companion) for Mathesar. The goal is UI-based upgrading of Mathesar when deployed via Docker Compose.

## In more detail

Below is a brief summary of Wiki.js Update Companion's README.

Implement a companion docker container that exposes a HTTP endpoint (`/upgrade`) to our main web service container (not the host), which, when queried will:

1. Stop and remove the Mathesar web service container
2. Pull the latest Mathesar web service image
3. Recreate the Mathesar container using the same settings as the previous container

```[tasklist]
## Tasks 
- [x] Fork Wiki.js Update Companion ([link to fork](https://github.com/centerofci/mathesar-update-companion))
- [x] #2194
- [ ] #2198
- [ ] #2195
- [ ] #2196
- [ ] #2197
- [ ] https://github.com/centerofci/mathesar/issues/2394
```

**Comment by dmos62** [comment]
## Our plan for this cycle

The install/upgrade process we're optimizing in this cycle, as I understand it, is git + `docker-compose up`.

However, if I understand correctly, the idea in this issue is to create a script that does this:

1. kill the container;
2. fetch latest published docker image;
3. build a fresh container out of that image.

Seems out of scope for this cycle, in that this is optimizing for a workflow where the user is running off of published docker images, instead of our git repo.

In this vein, we deprioritized #2040. This probably should be deprioritized for the same reason. What do you think @silentninja?

**Comment by silentninja** [comment]
> docker-compose up already does this, correct me if I'm wrong

Our production docker services won't be hot-loading files. 

>  optimizing for a workflow where the user is running off of published docker images, instead of our git repo

This issue makes the most sense for a pre-built docker image. We can do the same for the git repo workflow too by calling git pull and running the docker build command but it seems quite complicated and unnecessary. We skip this for now and work on it it when we publish a pre-built docker image.

**Comment by silentninja** [comment]
@dmos62 I re-evaluated the upgrade process for git-based flow and it isn't as bad as I thought it would be. I would prefer to do it in this cycle or the next cycle related to upgrading improvements as it provides a really wonderful UX.

@kgodey Which milestone should this go in if we don't have enough time to complete it in this cycle?

**Comment by kgodey** [comment]
@silentninja it should go in "2023-01 Public Launch"

**Comment by silentninja** [comment]
> By the way, is the goal here a companion container that would provide an HTTP endpoint (like the wiki.js example in the top post), or are we talking about a CLI script?

It will be an HTTP endpoint that can be triggered by the frontend.

**Comment by kgodey** [comment]
Sorry, I didn't see what was being discussed earlier, I only saw what I was tagged in.

> In this vein, we deprioritized https://github.com/centerofci/mathesar/issues/2040. This probably should be deprioritized for the same reason.

I had assumed that #2040 was de-prioritized because we can publish Docker releases manually for the first couple of releases instead of using CI. We should be supporting Docker-image based deployments for the initial launch. 

Providing users with as many easy installation and upgrade options as we can is critical to helping users adopt Mathesar.

**Comment by dmos62** [comment]
@kgodey I'll respond on Matrix.

@silentninja I'm a bit confused by this task and its relation to #2137. Is this task about creating the companion container and #2137 about giving it an `/upgrade` endpoint? Should they be a single task? Can you describe what an implementation of this task could be, and the same for #2137?

**Comment by silentninja** [comment]
>  Is this task about creating the companion container

Yes, this issue is about creating the script to replace the running docker container with the latest version. 

>https://github.com/centerofci/mathesar/issues/2137 about giving it an /upgrade endpoint?

Yes, #2137 is about having an endpoint to trigger the above script.

>Should they be a single task? 

They should be a single task, I made another one just to make it clear that there should be endpoint to trigger an upgrade. Please feel free to close #2137 and add its description this issue #2072 .

**Comment by dmos62** [comment]
Ok. I rewrote the top level post to just say that we want to implement the equivalent of the Wiki.js Update Companion. @silentninja provide feedback if my description isn't accurate.

**Comment by dmos62** [comment]
Added task list to top post.

**Comment by mathemancer** [comment]
Closed with wrong setting previously.
